### PR TITLE
Add option to disable MC labels in TRD TRAP sim

### PR DESCRIPTION
**Pull request by martenole**
This adds an option to the TRD TRAP simulator to disable MC label processing.
Unfortunately, it does not work anymore when labels are included. I get a segfault when I try to add the labels from the digits to the label container of the tracklets.
Instead of working directly with the input container 
auto lblDigits = pc.inputs().get<o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>>("labelinput")
I use a pointer to that container that is set to zero in case the label processing is disabled. I don't understand the error message why it fails to add an element to the MCTruthContainer. The relevant lines are:

[200135:TRAP]: #7  0x00007effedba6a37 in std::vector<o2::MCCompLabel, std::allocator<o2::MCCompLabel> >::emplace_back<o2::MCCompLabel const&> (this=<optimized out>) at /usr/include/c++/9/bits/alloc_traits.h:482
[200135:TRAP]: #8  o2::dataformats::MCTruthContainer<o2::MCCompLabel>::addElement (element=..., dataindex=1, this=0x7ffefee4eb20) at /home/oschmidt/alice/sw/SOURCES/O2/trapsim-nolabels/0/DataFormats/simulation/include/SimulationDataFormat/MCTruthContainer.h:208
[200135:TRAP]: #9  o2::dataformats::MCTruthContainer<o2::MCCompLabel>::addElements<o2::MCCompLabel const> (elements=..., dataindex=1, this=0x7ffefee4eb20) at /home/oschmidt/alice/sw/SOURCES/O2/trapsim-nolabels/0/DataFormats/simulation/include/SimulationDataFormat/MCTruthContainer.h:221
[200135:TRAP]: #10 o2::trd::TRDDPLTrapSimulatorTask::processTRAPchips (this=0x7effe38c2020, currDetector=<optimized out>, nTrackletsInTrigRec=
[200135:TRAP]: 0x7ffefee4eac8: 2, trapTrackletsAccum=std::vector of length 1, capacity 1 = {...}, lblTracklets=..., lblDigits=0x7ffefee4eae0) at /home/oschmidt/alice/sw/SOURCES/O2/trapsim-nolabels/0/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:173
[200135:TRAP]: #11 0x00007effedba993d in o2::trd::TRDDPLTrapSimulatorTask::run (this=0x7effe38c2020, pc=...) at /home/oschmidt/alice/sw/SOURCES/O2/trapsim-nolabels/0/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:277

Does anyone maybe have an idea what might go wrong?

### CODE REVIEW

File: Detectors/TRD/workflow/include/TRDWorkflow/TRDTrapSimulatorSpec.h [hunk 1]
```diff
@@ -57,10 +58,10 @@ class TRDDPLTrapSimulatorTask : public o2::framework::Task
   void loadTrapConfig();
   void loadDefaultTrapConfig();
   void setOnlineGainTables();
-  void processTRAPchips(int currDetector, int& nTrackletsInTrigRec, std::vector<Tracklet64>& trapTrackletsAccum, o2::dataformats::MCTruthContainer<o2::MCCompLabel>& trackletMCLabels, const o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>& digitMCLabels);
+  void processTRAPchips(int currDetector, int& nTrackletsInTrigRec, std::vector<Tracklet64>& trapTrackletsAccum, o2::dataformats::MCTruthContainer<o2::MCCompLabel>& lblTracklets, const o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>* lblDigits);
```
Comment (davidrohr) [inline Detectors/TRD/workflow/include/TRDWorkflow/TRDTrapSimulatorSpec.h:61]: shouldn't lblTracklets become a pointer as well?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 2]
```diff
@@ -221,25 +223,30 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
   }
+  LOG(debug) << "Trigger records are available for " << inputTriggerRecords.size() << " collisions";
 
   // output
   std::vector<Tracklet64> trapTrackletsAccum;                          // calculated tracklets
-  o2::dataformats::MCTruthContainer<o2::MCCompLabel> trackletMCLabels; // MC labels for the tracklets, taken from the digits which make up the tracklet (duplicates are removed)
   // copy from the input to keep the collision times, but the number of objects in here will refer to tracklets instead of digits
   std::vector<o2::trd::TriggerRecord> trackletTriggerRecords(inputTriggerRecords.begin(), inputTriggerRecords.end()); // time and number of tracklets for each collision
-  trapTrackletsAccum.reserve(inputDigits.size() / 500);
+  o2::dataformats::MCTruthContainer<o2::MCCompLabel> lblTracklets;                                                    // MC labels for the tracklets, taken from the digits which make up the tracklet (duplicates are removed)
```
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:252]: OK, I see, you create the container in any case, so with `--disable-mc` it just stays empty and is not shipped, should be ok then.

**Review by davidrohr — COMMENTED** [review]
I have to admit by only looking at the code I don't understand why it crashes if it worked before. Would have to debug it. If you don't make progress, you can ping me and I'll have a look.

**Comment by martenole** [comment]
Hi @sawenzel ,
do you maybe have an idea why the processing of the MC labels fails? Before this PR it worked fine. Now I added a switch whether or not the labels should be processed and don't use the label container object `o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>` directly, but a pointer to it. And as long as I didn't overlook something, this is the only change in the way I access the TRD digit labels. But now I get the error message above and don't know how to debug further.
Cheers,
Ole

**Comment by martenole** [comment]
I am still not making progress here. Am I missing something obvious? With the latest commit I simply revert back to using the container directly instead of the pointer to it. Nothing else is changed. Now everything works fine.
But when I try to work with the pointer I get the above described segfault.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 3]
```diff
@@ -221,25 +223,30 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
   if (inputDigits.size() == 0 || inputTriggerRecords.size() == 0) {
     LOG(warn) << "Did not receive any digits, trigger records, or neither one nor the other. Aborting.";
     return;
   }
   LOG(debug) << "Read in " << inputDigits.size() << " digits";
-  LOG(debug) << "Labels contain " << digitMCLabels.getNElements() << " elements with and indexed size of " << digitMCLabels.getIndexedSize();
+
+  const o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>* lblDigitsPtr = nullptr;
+
+//  if (mUseMC) {
```
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: The problem is here, `auto lblDigits` goes out of scope after the `if (museMC) { ... }` branch, so when lblDigitsPtr is used below, it is invalid.

**Review by davidrohr — CHANGES_REQUESTED** [review]
@martenole : Applying this patch on your first PR should fix it:
```
diff --git a/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx b/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
index 8623f33f4fc..a1be85d49d7 100644
--- a/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
+++ b/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
@@ -11,6 +11,7 @@
 #include "TRDWorkflow/TRDTrapSimulatorSpec.h"
 
 #include <chrono>
+#include <optional>
 #include <gsl/span>
 
 #include "TFile.h"
@@ -231,10 +232,12 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
   LOG(debug) << "Read in " << inputDigits.size() << " digits";
 
   const o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>* lblDigitsPtr = nullptr;
+  using lblType = std::decay_t<decltype(pc.inputs().get<o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>>(""))>;
+  std::optional<lblType> lblDigits;
 
   if (mUseMC) {
-    auto lblDigits = pc.inputs().get<o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>>("labelinput"); // MC labels associated to the input digits
-    lblDigitsPtr = &lblDigits;
+    lblDigits.emplace(pc.inputs().get<o2::dataformats::ConstMCTruthContainer<o2::MCCompLabel>>("labelinput")); // MC labels associated to the input digits
+    lblDigitsPtr = &lblDigits.value();
     LOG(debug) << "Labels contain " << lblDigitsPtr->getNElements() << " elements with and indexed size of " << lblDigitsPtr->getIndexedSize();
     if (lblDigitsPtr->getIndexedSize() != inputDigits.size()) {
       LOG(warn) << "Digits and Labels coming into TrapSimulator are of differing sizes, labels will be jibberish. " << lblDigitsPtr->getIndexedSize() << "!=" << inputDigits.size();
```
The std::optional will store the ConstMCTruthContainer class, only in case we receive the labels. Would probably also work with a std::move, but this way we ensure there can never by any copying.

**Comment by davidrohr** [comment]
Could you please also add the `GetUncalibratedY` and `GetUncalibratedDy` getters to the AliRoot data type.

**Comment by davidrohr** [comment]
> Could you please also add the `GetUncalibratedY` and `GetUncalibratedDy` getters to the AliRoot data type.

**Please ignore this! I had a problem compiling the standalone benchmark, but for a different reason...**

But I would move the getters to the tracklet64 header.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 3 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Thank you very much, that makes perfectly sense. I just didn't see it! After the compilation is done and I have tested it I will push the changes.

**Comment by martenole** [comment]
Why/how should I move the getters to the Tracklet64 header? You mean from Tracklet64.cxx? In principle the GPUTRDTrackletWord is an interface to either AliRoot or o2 type tracklets. GetY() and GetdY() are the methods which are used in the code. For o2 they point to the Tracklet64::getUncalibratedY() and getUncalibratedDy()

**Comment by davidrohr** [comment]
Yes, it was my fault. Compilation is fine. I would just move the gtters from the cxx to the h to avoid the function call. I'd place them below the class definition with GPUdi() instead of GPUd.

**Comment by davidrohr** [comment]
[inline.patch.gz](https://github.com/AliceO2Group/AliceO2/files/6132338/inline.patch.gz)
this is what I meant.

**Comment by martenole** [comment]
Thanks, I probably would have just put it in the class definition directly

**Comment by martenole** [comment]
Once the checks are through this is ready to be merged

**Pull request merged**